Mathlib/CategoryTheory/Shift/Twist.lean

```diff
@@ -0,0 +1,119 @@
+category `TwistShift t` which identical shift functors as `C` but where
+the isomorphisms `shiftFunctorAdd` have been modified.
+
+-/
+
+@[expose] public section
+
+universe w v u
+
+namespace CategoryTheory
+
+variable (C : Type u) [Category.{v} C] (A : Type w) [AddMonoid A] [HasShift C A]
+
+/-- Given a category `C` equipped with a shift by a monoid `A` -/
+structure TwistShiftData where
+  /-- The invertible elements in the center of `C` which are used to
+  modify the `shiftFunctorAdd` isomorphisms . -/
+  z (a b : A) : (CatCenter C)ˣ
+  z_zero_right (a : A) : z a 0 = 1 := by cat_disch
+  z_zero_left (b : A) : z 0 b = 1 := by cat_disch
+  assoc (a b c : A) : z (a + b) c * z a b = z a (b + c) * z b c := by cat_disch
+  commShift (a b : A) : NatTrans.CommShift (z a b).val A := by infer_instance
+
+namespace TwistShiftData
+
+variable {C A} (t : TwistShiftData C A)
+
+attribute [simp] z_zero_right z_zero_left
+attribute [instance] commShift
+
+@[simp]
+lemma shift_z_app (a b c : A) (X : C) :
+    ((t.z a b).val.app X)⟦c⟧' = (t.z a b).val.app (X⟦c⟧) := by
+  simpa using NatTrans.shift_app_comm (t.z a b).val c X
+
+end TwistShiftData
+
+variable {C A}
+
+/-- Given `t : TwistShiftData C A`, this is a type synonym for the category `C`,
+which the same shift functors as `C` but where the `shiftFunctorAdd` isomorphisms
+have been modified using `t`. -/
+@[nolint unusedArguments]
+def TwistShift (_ : TwistShiftData C A) : Type u := C
+
+variable (t : TwistShiftData C A)
+
+namespace TwistShift
+
+instance : Category (TwistShift t) := inferInstanceAs (Category C)
+
+/-- Given `t : TwistShiftData C A`, the shift on the category `TwistShift t` has
+the same shift functors as `C`, the same isomorphism `shiftFunctorZero` isomorphism,
+but the `shiftFunctorAdd` isomorphisms are modified using `t`. -/
+def shiftMkCore : ShiftMkCore (TwistShift t) A where
+  F a := shiftFunctor C a
+  zero := shiftFunctorZero C A
+  add a b := NatIso.ofComponents (fun X ↦ t.z a b • (shiftFunctorAdd C a b).app X) (by
+    simp [CatCenter.naturality_assoc, CatCenter.naturality])
+  add_zero_hom_app := by simp [shiftFunctorAdd_add_zero_hom_app]
+  zero_add_hom_app := by simp [shiftFunctorAdd_zero_add_hom_app]
+  assoc_hom_app a b c X := by
+    dsimp
+    simp only [Functor.map_comp, Category.assoc]
+    rw [CatCenter.naturality, CatCenter.naturality_assoc, CatCenter.naturality_assoc,
+      CatCenter.naturality_assoc, CatCenter.naturality_assoc, CatCenter.naturality_assoc,
+      t.shift_z_app, CatCenter.naturality, CatCenter.naturality_assoc,
+      ← CatCenter.mul_app_assoc, ← CatCenter.mul_app_assoc,
+      ← Units.val_mul, ← Units.val_mul, t.assoc a b c]
+    simp [shiftFunctorAdd_assoc_hom_app (C := C) a b c X, shiftFunctorAdd']
+
+instance : HasShift (TwistShift t) A := hasShiftMk _ _ (shiftMkCore t)
+
+@[simp]
+lemma shiftFunctor_obj (X : TwistShift t) (a : A) :
+    (shiftFunctor (TwistShift t) a).obj X = (shiftFunctor C a).obj X := rfl
```

Comment:
The two sides of this equation are not reducibly equal. This might cause problems downstream?

Reply:
It will not change much for the downstream applications, but I have now put more API isolation.